param.go

```diff
@@ -105,6 +115,50 @@ func (f paramVisitorFunc) Visit(p param) paramVisitor {
 	return nil
 }
 
+type paramVisitorSetter interface {
+	With(paramVisitorFunc) paramVisitor
+}
+
+// paramVisitOnce encapsulates a paramVisitorFunc and a map to track visited nodes.
+type paramVisitOnce struct {
+	f       paramVisitorFunc
+	visited map[key]struct{}
+}
+
+// NewParamVisitOnce constructs a paramVisitor avoiding redundant visits.
+func newParamVisitOnce() paramVisitorSetter {
+	return paramVisitOnce{
+		visited: make(map[key]struct{}),
+	}
+}
+
+func (pv paramVisitOnce) With(f paramVisitorFunc) paramVisitor {
+	return paramVisitOnce{
```

Comment:
Why make this a two-part constructor instead of
`newParamVisitOnce(paramVisitor) paramVisitor`?